src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,168 @@
+
     @Override
     public void add(T value) {
+
+        Node<T> newNode = new Node<>(value, tail, null);
+        if (tail == null) {
+            head = newNode;
+        } else {
+            tail.next = newNode;
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> nextNode = getNodeByIndex(index);
+            Node<T> prevNode = nextNode.prev;
+            Node<T> newNode = new Node<>(value, prevNode, nextNode);
+
+            if (prevNode == null) {
+                head = newNode;
+            } else {
+                prevNode.next = newNode;
+            }
+            nextNode.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNodeByIndex(index).data;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNodeByIndex(index);
+        T oldValue = node.data;
+        node.data = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> nodeToRemove = getNodeByIndex(index);
+        return unlink(nodeToRemove);
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        if (object == null) {
+            while (current != null) {
+                if (current.data == null) {
+                    unlink(current);
+                    return true;
+                }
+                current = current.next;
+            }
+        } else {
+            while (current != null) {
+                if (object.equals(current.data)) {
+                    unlink(current);
+                    return true;
+                }
+                current = current.next;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private T unlink(Node<T> node) {
+        final T element = node.data;
+        Node<T> prev = node.prev;
+        Node<T> next = node.next;
+
+        if (prev == null) {
+            head = next;
+        } else {
+            prev.next = next;
+            node.prev = null;
+        }
+
+        if (next == null) {
+            tail = prev;
+        } else {
+            next.prev = prev;
+            node.next = null;
+        }
+
+        node.data = null;
+        size--;
+        return element;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
```

Comment:
Don't create repeating code. Consider creating a `findNodeByIndex` method to avoid code duplication in methods like `get`, `set`, `remove(int index)`, and `add(T value, int index)`. This method is already implemented here as `getNodeByIndex`, so ensure it's used consistently to avoid duplication.